### PR TITLE
feat: native wallet passphrase recovery

### LINKED ISSUE #2427: Allow Native Wallet users to back up their wallets
### Overview

Users using the Native wallet inside [app.shapeshift.com](http://app.shapeshift.com/) currently can only access their seed phrase after login into the Native wallet. Additionally, users who import their seed phrase from another wallet into a Native wallet, don’t have the ability to access their seed phrase anywhere in the web app again while using their imported wallet. The goal of this feature is to give users on Native wallet the ability to “back up their wallet” (while using their wallet).

### References and additional details

Please reference this spec for the project scope. 
https://shapeshift.notion.site/Back-Up-Native-Wallet-4387641bedcb4bf3a9d2d24096f6c5ae

### Acceptance Criteria

Users should be able access their Native wallet seed phrase, while using their Native Wallet
User should be able to test that they have written down their seed phrase correctly.

### Need By Date

Aug 18

### Screenshots/Mockups

Here is a link to the flow (it should use most of what we already have from the native wallet migration:
https://www.figma.com/proto/eSaTFovDi9Re3ELrK2npkt/Back-Up-Wallet?node-id=1501%3A5630&scaling=min-zoom&page-id=1501%3A5628

### Estimated effort

3 days?

### PR DESCRIPTION
**Pull request by stackedq**
## Description

Adds the functionality to "backup your passphrase" for ShapeShift native wallet.

Also refactors how `ConnectedWalletMenu` show related items for the connected wallet.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #2427 

## Risk

none

## Testing

1. connect ShapeShift Native wallet
2. Inside the wallet dropdown in the upper right corner, users will see a new option “Back up my wallet”
3. This will open a modal that asks for the password, and then the user will be presented with an intro screen, that warns them to write down their secret recovery phrase and store it in a safe place.
4. Users will be taken through the same flow that exists for importing a wallet, ie, They will see the screen where they can write down their secret recovery phrase 
5. Users will see 3 subsequent screens to test their seed phrase, or they have the option to check a box, indicating they already have it written down successfully.
6. a success message will be shown afterward

## Screenshots (if applicable)
Screen Shot 2022-08-18 at 13 39 39
Screen Shot 2022-08-18 at 13 39 48
Screen Shot 2022-08-18 at 13 40 01
Screen Shot 2022-08-18 at 13 40 35
Screen Shot 2022-08-18 at 13 40 43
Screen Shot 2022-08-18 at 13 40 52
Screen Shot 2022-08-18 at 13 41 00
Screen Shot 2022-08-18 at 13 41 10

### CODE REVIEW

File: src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseInfo.tsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    ;(async () => {
+      try {
+        setWords(
+          (await vault.unwrap().get('#mnemonic')).split(' ').map((word: string, index: number) => (
+            <Tag
+              p={2}
+              flexBasis='31%'
+              justifyContent='flex-start'
+              fontSize='md'
+              key={word}
```
Comment (gomesalexandre) [inline src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseInfo.tsx:53]: Usually an antipattern, but I'd rather use the index as a key here for security reasons. At this point the words are already in clear in the browser, and rendered as a children but still, the less places they are in the better

File: src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseTest.tsx [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+export const BackupPassphraseTest = ({ vault }: { vault: Vault | null }) => {
+  const translate = useTranslate()
+  const history = useHistory()
+  const [testState, setTestState] = useState<TestState | null>(null)
+  const [invalidTries, setInvalidTries] = useState<number[]>([])
+  const [testCount, setTestCount] = useState<number>(0)
+  const [revoker] = useState(new (Revocable(class {}))())
+  const [shuffledNumbers] = useState(slice(shuffle(range(12)), 0, TEST_COUNT_REQUIRED))
```
Comment (gomesalexandre) [inline src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseTest.tsx:48]: Abstract away to `initialShuffledNumbersState` and/or disable eslint for it? Clojure-style expressions are hard to read when one-lined

File: src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseTest.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+import { RawText, Text } from 'components/Text'
+
+import { BackupPassphraseRoutes } from './BackupPassphraseCommon'
+
+const Revocable = native.crypto.Isolation.Engines.Default.Revocable
+const revocable = native.crypto.Isolation.Engines.Default.revocable
+
+const TEST_COUNT_REQUIRED = 3
+
+const ordinalSuffix = (n: number) => {
```
Comment (gomesalexandre) [inline src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseTest.tsx:31]: I know this is taken from a current implementation, but let's not spread this naming and make it clear this is a function and not an expression
```suggestion
const makeOrdinalSuffix = (n: number) => {
```

**Review by gomesalexandre — COMMENTED** [review]
Tested and while this mostly looks fine functionally, noticed these discrepancies with the spec:

- `The user must pick from a selection of 15 words. If it’s correct, the user is moved forward to the next screen. If its incorrect, the screen re-scrambles and asks the user again.` - this doesn't re-scramble and it is possible to select from any of the scrambled words, until the right one is eventually chosen
- `User will see 3 subsequent screens to test their seed phrase, or they have the option to check a box, indicating they already have it written down successfully.` - this one isn't present in the PR

Are those approved by product as a deviation from the initial spec?

File: src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseInfo.tsx (hunk 1 above)
Comment (0xdef1cafe) [inline src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseInfo.tsx:53]: just for safety make this the index

**Comment by stackedq** [comment]
Reshuffling on failure and skip added
Screen Shot 2022-08-19 at 09 53 40

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
@stackedq looking good with the checkbox now, and the reshuffling works but with it implemented perhaps we should also remove the "already tried" logic/UI, WDYT?

Also with the current logic, it is possible to end up in a state where all wrong tries are used, and it is not possible to select the right word anymore, see:

image

**Review by gomesalexandre — APPROVED** [review]
LGTM now with the removal of the "invalid tries" flow, retested the whole flow as a sanity check and does what it says on the box.

Well done ser! 🤵‍♂️

**Pull request merged**